base/ltluatex.dtx

```diff
@@ -1284,14 +1284,18 @@ luatexbase.new_chunkname = new_chunkname
 %
 % \begin{macro}{new_luafunction}
 % \changes{v1.1i}{2018/10/21}{Function added}
+% \changes{v1.2c}{2023/07/02}{Ensure existing table entries not overwritten gh/1100}
 % Much the same as for attribute allocation in Lua.
 % The optional \meta{name} argument is used in the log if given.
 %    \begin{macrocode}
 local luafunction_count_name =
                          luafunction_count_name or "e@alloc@luafunction@count"
 local function new_luafunction(name)
   tex_setcount("global", luafunction_count_name,
-                         tex_count[luafunction_count_name] + 1)
+                         math.max(
+                           #(lua.get_functions_table()),
+                           tex_count[luafunction_count_name])
+                          + 1)
```

Comment:
We should additionally do something like

```
lua.get_functions_table()[tex_count[luafunction_count_name]] = false
```

to ensure that code relying on the table length doesn't consider a slot to be free after it has been returned by `new_luafunction` in case it's not immediately used.

Reply:
@zauguin sounds good (also highlights we shoudl probably never had had a count on the tex side at all, but since we do, ensuring it really is the same as the Lua table size sounds like a good plan), I'll add to this branch

Reply:
resolved by fbac7f05